### PR TITLE
plain write: fixed bug with incorrect footer size computing. It happe…

### PR DESCRIPTION
**Pull request by abudnik**
…ns on rewriting of a key that have empty flags to a key with NOCSUM flags in eblob_plain_writev_prepare()

### CODE REVIEW

File: library/blob.c [hunk 1]
```diff
@@ -330,10 +330,12 @@ inline static uint64_t eblob_validate_ctl_flags(struct eblob_backend *b, uint64_
...
-	if (!(flags & BLOB_DISK_CTL_NOCSUM))
-		flags |= BLOB_DISK_CTL_CHUNKED_CSUM;
-	else
-		flags &= ~BLOB_DISK_CTL_CHUNKED_CSUM;
+	/*
+	 * We have to add BLOB_DISK_CTL_CHUNKED_CSUM because it shows
+	 * that footer was prepared with specific size, even if the record
+	 * has BLOB_DISK_CTL_NOCSUM flag.
+	 */
+	flags |= BLOB_DISK_CTL_CHUNKED_CSUM;
```
Comment (bioothod) [inline library/blob.c]: Please elaborate why is this needed?
This looks like a but - we always set flag even if we do not want checksum.

Comment (shaitan) [inline library/blob.c]: It is needed because even if a record wasn't checksummed (because it was written with `BLOB_DISK_CTL_NOCSUM`), it still has footer whose size is different for old and for new data.

Flag `BLOB_DISK_CTL_CHUNKED_CSUM` shows that footer has specific structure. 
Flag `BLOB_DISK_CTL_NOCSUM` shows that checksums from the footer is unfilled and should be ignored.

We always set flag `BLOB_DISK_CTL_CHUNKED_CSUM` to new record's flag because we do not write records with old checksums anymore.

Comment (bioothod) [inline library/blob.c]: Does it mean we always set this flag in the header and always allocate space for the footer, but if checksum was disabled, footer will be empty?

If that is true, how does it know whether to perform checksum verification or not?

Comment (shaitan) [inline library/blob.c]: > Does it mean we always set this flag in the header and always allocate space for the footer, but if checksum was disabled, footer will be empty?
Yes.

>If that is true, how does it know whether to perform checksum verification or not?
It checks flags and if requested record has `BLOB_DISK_CTL_NOCSUM` it won't perform checksum verification.

Comment (shaitan) [inline library/blob.c]: Also if eblob was configure with `EBLOB_NO_FOOTER` all records will be written without any footer, so eblob won't perform checksum verification.

Comment (bioothod) [inline library/blob.c]: I meant different thing.
Let's suppose we have a record with BLOB_DISK_CTL_CHUNKED_CSUM bit set, it also has a footer, but it is empty.
Previously it was enough just to look at record's flag to determine that there is no checksum and thus eblob should not try to verify it. We plan to always have BLOB_DISK_CTL_CHUNKED_CSUM bit now, and how should eblob make verification decision? Looks like it will have to go to the footer, calculate its size, determine that it is empty, and only then decide not to verify checksum.

What's wrong with *not* setting BLOB_DISK_CTL_CHUNKED_CSUM bit if we do not actually plan to put checksum into footer?

Comment (shaitan) [inline library/blob.c]: >I meant different thing.
Let's suppose we have a record with BLOB_DISK_CTL_CHUNKED_CSUM bit set, it also has a footer, but it is empty.
Previously it was enough just to look at record's flag to determine that there is no checksum and thus eblob should not try to verify it.

It continues to be true with one change: it is enough to look at record's flag and if `BLOB_DISK_CTL_NOCSUM` is set then checksum is empty and it shouldn't be verified.

>What's wrong with not setting BLOB_DISK_CTL_CHUNKED_CSUM bit if we do not actually plan to put checksum into footer?

If eblob wasn't configured with `EBLOB_NO_FOOTER` then all records have footer. Footer size is different for records with old and with new checksums. Without `BLOB_DISK_CTL_CHUNKED_CSUM` we can't determine which checksums (new or old) are contained in the footer, so we can't calculate its size.

The presence or absence of `BLOB_DISK_CTL_CHUNKED_CSUM` says nothing about that checksums are filled or not, so it shouldn't be used for such purpose.

Comment (bioothod) [inline library/blob.c]: Please add description of this logic and how each particular bitbehaves into `include/eblob/blob.h`

Comment (abudnik) [inline library/blob.c]: done, see https://github.com/abudnik/eblob/commit/22c9e4a2bff7498b09f7d732867713e58e9bd305

Comment (bioothod) [inline library/blob.c]: Thank you
